### PR TITLE
GoLang training: banking application

### PR DESCRIPTION
**Pull request by dominiclopes**

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -0,0 +1,56 @@
+module example.com/banking
```
Comment (sushmitaw) [inline go.mod]: example.com is not appropriate module name , generally we use github.com/username/banking

File: go.mod [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+	go.uber.org/atomic v1.7.0 // indirect
+	go.uber.org/multierr v1.6.0 // indirect
+	golang.org/x/sys v0.0.0-20220722155257-8c9f86f7a55f // indirect
+	golang.org/x/text v0.3.7 // indirect
+	gopkg.in/ini.v1 v1.67.0 // indirect
+	gopkg.in/yaml.v2 v2.4.0 // indirect
+	gopkg.in/yaml.v3 v3.0.1 // indirect
+)
+
+replace example.com/banking => /home/dominic/Documents/GoTrainingAtJosh/BankingApplication
```
Comment (sushmitaw) [inline go.mod]: replace directive should be used only in local , we do not have the destination path same on peer machine or prod

File: app/app.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+	logger = zapLogger.Sugar()
+}
+
+func initDB() (err error) {
+
+	dbConfig := config.Database()
+
+	db, err = sqlx.Open(dbConfig.Driver(), dbConfig.ConnectionURL())
+	if err != nil {
+		return
```
Comment (sushmitaw) [inline app/app.go]: add error log

File: app/app.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+	}
+
+	if err = db.Ping(); err != nil {
+		return
```
Comment (sushmitaw) [inline app/app.go]: add error log

File: api/responses.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+	"net/http"
+
+	"example.com/banking/app"
+)
+
+type Response struct {
+	Message string `json:"message"`
+}
+
+func Error(rw http.ResponseWriter, status int, response interface{}) {
```
Comment (sushmitaw) [inline api/responses.go]: there is no difference between Success & error , when returning error we should send error msg , custom error code if neccessary

File: bank/handler.go [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		var uAuth LoginRequest
+
+		err := json.NewDecoder(req.Body).Decode(&uAuth)
+		if err != nil {
+			api.Error(rw, http.StatusBadRequest, api.Response{Message: err.Error()})
+			return
+		}
+		// Validate if the user email and phonenumber is correct
+		if uAuth.Email == "" || uAuth.Password == "" {
```
Comment (sushmitaw) [inline bank/handler.go]: create a new receiver method (l *LoginRequest) Validate() error {} put all the logic in thr .....

File: bank/handler.go [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+		if err != nil {
+			if err == ErrUnauthorized {
+				api.Error(rw, http.StatusUnauthorized, api.Response{Message: "Unauthorized"})
+				return
+			}
+			api.Error(rw, http.StatusInternalServerError, api.Response{Message: "Err - Internal Server Error"})
+			return
+		}
+
+		http.SetCookie(rw, &http.Cookie{
```
Comment (sushmitaw) [inline bank/handler.go]: generally we use authorization header to accept/set token , rather than cookie

File: bank/handler.go [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+		var accReq CreateAccountRequest
+
+		err = json.NewDecoder(req.Body).Decode(&accReq)
+		if err != nil {
+			api.Error(rw, http.StatusInternalServerError, api.Response{Message: "Internal Server Error"})
+			return
+		}
+
+		// Validate if the user email and phonenumber is correct
+		if accReq.Email == "" || accReq.PhoneNumber == "" {
```
Comment (sushmitaw) [inline bank/handler.go]: same as above add validation method

File: bank/handler.go [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+
+		var transactionDetailsRequest GetTransactionDetailsRequest
+		err = json.NewDecoder(req.Body).Decode(&transactionDetailsRequest)
+		if err != nil {
+			api.Error(rw, http.StatusBadRequest, api.Response{Message: err.Error()})
+			return
+		}
+
+		// Validate the formats of the start and end date
+		startDateTime, err := time.Parse("2006-01-02", transactionDetailsRequest.StartDate)
```
Comment (sushmitaw) [inline bank/handler.go]: same as above

File: bank/service.go [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+}
+
+func NewBankService(s db.Storer, l *zap.SugaredLogger) Service {
+	return &bankService{
+		store:  s,
+		logger: l,
+	}
+}
+
+func generateJWT(userID string, role string) (tokenString string, tokenExpirationTime time.Time, err error) {
```
Comment (sushmitaw) [inline bank/service.go]: add all jwt methods to utls

File: bank/service.go [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
+	tokenString, err = token.SignedString(secretKey)
+	if err != nil {
+		err = fmt.Errorf("error generating token, err: %v", err)
+		return
+	}
+	return
+}
+
+func ValidateJWT(tokenString string) (claims *Claims, err error) {
```
Comment (sushmitaw) [inline bank/service.go]: add all jwt methods to utls

File: bank/service.go [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+	if err != nil {
+		err = fmt.Errorf("unauthorized, err: %v", err)
+		return
+	}
+	return
+}
+
+func (b *bankService) Login(ctx context.Context, u LoginRequest) (tokenString string, tokenExpirationTime time.Time, err error) {
+	// Verify if user is present
+	user, err := b.store.GetUserByEmailAndPassword(ctx, u.Email, u.Password)
```
Comment (sushmitaw) [inline bank/service.go:38]: save all personal data in encrypted form in db

File: bank/service.go [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+	return
+}
+
+func (b *bankService) CreateAccount(ctx context.Context, accReq CreateAccountRequest) (accRes CreateAccountResponse, err error) {
+	b.logger.Infof("Creating an account for user email: %v, phone number: %v\n", accReq.Email, accReq.PhoneNumber)
+
+	// Create the user ID, password and update the balance
+	u := db.User{
+		Email:       accReq.Email,
+		PhoneNumber: accReq.PhoneNumber,
```
Comment (sushmitaw) [inline bank/service.go:66]: same as above

**Review by sushmitaw — CHANGES_REQUESTED** [review]

File: go.mod (hunk 2 above)
Comment (dominiclopes) [reply go.mod]: As I was using an incorrect module name, I had used the replace directive.
Removed this as we have updated for using the correct module name

File: app/app.go (hunk 3 above)
Comment (dominiclopes) [reply app/app.go]: Updated, using logrus for logging the error and exiting the program. Thus, skipping the use of return

File: app/app.go (hunk 4 above)
Comment (dominiclopes) [reply app/app.go]: Updated, using logrus for logging the error and exiting the program. Thus, skipping the use of return

File: go.mod (hunk 1 above)
Comment (dominiclopes) [reply go.mod]: Updated the module name to `github.com/dominiclopes/BankingApplication`

File: api/responses.go (hunk 5 above)
Comment (dominiclopes) [reply api/responses.go]: Remove the duplicate code and updated the function name to APIResponse

File: bank/handler.go (hunk 6 above)
Comment (dominiclopes) [reply bank/handler.go]: Updated, added the validations in a method, and using the same in the handler

File: bank/handler.go (hunk 9 above)
Comment (dominiclopes) [reply bank/handler.go]: Added the validations in a method

File: bank/service.go (hunk 10 above)
Comment (dominiclopes) [reply bank/service.go]: moved the jwt generate and validate functions to utils>jwt.go

File: bank/service.go (hunk 11 above)
Comment (dominiclopes) [reply bank/service.go]: moved the jwt generate and validate functions to utils>jwt.go

File: bank/handler.go (hunk 7 above)
Comment (dominiclopes) [reply bank/handler.go]: Updated - getting the token from the header instead of cookie

**Pull request merged**